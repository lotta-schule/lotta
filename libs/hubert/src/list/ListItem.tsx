import * as React from 'react';
import { Divider } from '../divider/Divider';
import clsx from 'clsx';

import styles from './List.module.scss';

export type ListItemProps = React.HTMLProps<HTMLLIElement> & {
  leftSection?: React.ReactNode;
  rightSection?: React.ReactNode;
  isDivider?: boolean;
  isDisabled?: boolean;
  isHeader?: boolean;
  isFocused?: boolean;
  isSelected?: boolean;
};

export const ListItem = ({
  children,
  className,
  isDisabled,
  isFocused,
  isSelected,
  isHeader,
  leftSection,
  rightSection,
  ...props
}: ListItemProps) => {
  const internalRef = React.useRef<HTMLLIElement>(null);
  if (props.isDivider === true) {
    return (
      <li
        className={clsx(styles.li, styles.isDivider, className)}
        {...Object.fromEntries(
          Object.entries(props).filter(([key]) => key !== 'isDivider')
        )}
      >
        <Divider />
      </li>
    );
  }

  return (
    <li
      className={clsx(styles.li, className, {
        [styles.isClickable]: props.onClick !== undefined,
        [styles.isDisabled]: isDisabled,
        [styles.isFocused]: isFocused,
        [styles.isSelected]: isSelected,
        [styles.isHeader]: isHeader,
      })}
      {...(isFocused ? { ['data-is-focused']: true } : {})}
      {...props}
      ref={(node) => {
        if (props.ref) {
          if ('current' in props.ref) {
            props.ref.current = node;
          } else {
            props.ref(node);
          }
        }
        internalRef.current = node;
        return () => {
          if (props.ref) {
            if ('current' in props.ref) {
              props.ref.current = null;
            } else {
              props.ref(null);
            }
          }
          internalRef.current = null;
        };
      }}
    >
      {leftSection && <div>{leftSection}</div>}
      <div className={styles.mainSection}>{children}</div>
      {rightSection && <div>{rightSection}</div>}
    </li>
  );
};
ListItem.displayName = 'ListItem';
