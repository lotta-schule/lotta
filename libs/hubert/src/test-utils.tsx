import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, type RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';
import { BrowserNode, BrowserState, NodeList } from 'browser';
import { BrowserStateContext } from 'browser/BrowserStateContext';
import * as fixtures from 'test-fixtures';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement<any>,
  renderOptions: RenderOptions = {}
): ReturnType<typeof render> =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const createPromise = <T = void,>() => {
  let resolve: (value: T) => void;
  let reject: (error?: any) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve: resolve!, reject: reject! };
};

// For Testing components that use the BrowserStateContext
export type TestBrowserWrapperProps = {
  children?: React.ReactNode;
  defaultNodes?: BrowserNode[];
} & Partial<Omit<BrowserState, 'onRequestChildNodes' | 'renderNodeList'>>;
export const TestBrowserWrapper = ({
  children,
  mode = 'view-and-edit',
  defaultNodes = fixtures.browserNodes,
  moveNode = vi.fn(),
  renameNode = vi.fn(),
  deleteNode = vi.fn(),
  searchNodes = vi.fn(),
  currentPath = [],
  selected = [],
  isNodeDisabled = vi.fn(() => false),
  onSelect = vi.fn(),
  onNavigate = vi.fn(),
  setCurrentAction = vi.fn(),
  currentAction = null,
  currentSearchResults = null,
  setCurrentSearchResults = vi.fn(),
  canEdit,
  uploadClient = {
    addFile: vi.fn(),
    byState: { uploading: [], pending: [], done: [], error: [] },
    hasErrors: false,
    isSuccess: false,
    isUploading: false,
    currentProgress: null,
    currentUploads: [],
  },
}: TestBrowserWrapperProps) => {
  const [nodes, setNodes] = React.useState<BrowserNode[]>(defaultNodes);
  const [isFilePreviewVisible, setIsFilePreviewVisible] = React.useState(false);

  const onRequestChildNodes = (node: BrowserNode<'directory'> | null) =>
    nodes.filter((n) => n.parent === (node?.id ?? null));

  return (
    <BrowserStateContext.Provider
      value={{
        uploadClient,
        canEdit,
        onRequestChildNodes: async (node) =>
          onRequestChildNodes(node as BrowserNode<'directory'>),
        renderNodeList: ({ path }) => (
          <NodeList path={path} nodes={onRequestChildNodes(path.at(-1)!)} />
        ),
        currentAction,
        onSelect,
        onNavigate,
        currentSearchResults,
        currentPath,
        selected,
        mode,
        isNodeDisabled,
        setCurrentAction,
        setCurrentSearchResults,
        moveNode,
        renameNode,
        deleteNode,
        searchNodes,
        isFilePreviewVisible,
        setIsFilePreviewVisible,
        createDirectory: async (parentNode, name) => {
          setNodes((nodes) => {
            return [
              ...nodes,
              {
                id: `${nodes.length + 1}`,
                name,
                type: 'directory',
                parent: parentNode?.id ?? null,
                meta: {},
              },
            ];
          });
        },
      }}
    >
      {children}
    </BrowserStateContext.Provider>
  );
};

export * as fixtures from './test-fixtures';
