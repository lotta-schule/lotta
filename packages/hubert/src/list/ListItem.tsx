import * as React from 'react';
import { Divider } from '../divider/Divider';
import clsx from 'clsx';

import styles from './List.module.scss';

export type ListItemProps = React.HTMLProps<HTMLLIElement> & {
  leftSection?: React.ReactNode;
  rightSection?: React.ReactNode;
  isDivider?: boolean;
  isDisabled?: boolean;
  isFocused?: boolean;
  isSelected?: boolean;
};

export const ListItem = React.forwardRef(
  (
    {
      children,
      className,
      isDisabled,
      isFocused,
      isSelected,
      leftSection,
      rightSection,
      ...props
    }: ListItemProps,
    ref: React.ForwardedRef<HTMLLIElement>
  ) => {
    if (props.isDivider === true) {
      return (
        <li
          className={clsx(styles.li, styles.isDivider, className)}
          {...Object.fromEntries(
            Object.entries(props).filter(([key]) => key !== 'isDivider')
          )}
        >
          <Divider />
        </li>
      );
    }

    return (
      <li
        className={clsx(styles.li, className, {
          [styles.isDisabled]: isDisabled,
          [styles.isFocused]: isFocused,
          [styles.isSelected]: isSelected,
        })}
        {...(isFocused ? { ['data-is-focused']: true } : {})}
        ref={ref}
        {...props}
      >
        {leftSection && <div>{leftSection}</div>}
        <div className={styles.mainSection}>{children}</div>
        {rightSection && <div>{rightSection}</div>}
      </li>
    );
  }
);
ListItem.displayName = 'ListItem';
