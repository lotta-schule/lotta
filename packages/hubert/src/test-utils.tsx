import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement,
  renderOptions: Omit<RenderOptions, 'wrapper'> = {}
) =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
