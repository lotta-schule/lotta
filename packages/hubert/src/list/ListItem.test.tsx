import * as React from 'react';
import { render } from '../test-utils';
import { ListItem } from './ListItem';

describe('list/ListItem', () => {
  it('should render a list item', () => {
    const screen = render(<ListItem>Item</ListItem>);
    expect(screen.getByRole('listitem')).toMatchSnapshot();
  });

  it('should show a divider when isDivider prop is passed', () => {
    const screen = render(<ListItem isDivider />);
    expect(screen.getByRole('separator')).toBeVisible();
  });
});
