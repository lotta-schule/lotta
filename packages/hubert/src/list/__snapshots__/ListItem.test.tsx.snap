// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list/ListItem should render a list item 1`] = `
<li
  class="li"
>
  <div
    class="mainSection"
  >
    Item
  </div>
</li>
`;
